KAFKA-3794: added stream / table names as prefix to functions that pr…

bbejeck opened the pull request:
…int to the console.

bbejeck committed:
KAFKA-3794: added stream / table names as prefix to functions that print to the console.

bbejeck commented:
@guozhangwang 
- Changed my position a bit on the `writeAsText`, files looked better (to me at least) without the label.
- KTables use the name provided at instantiation for the name on the console.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:
```diff
@@ -106,11 +119,12 @@
+     * type String, Integer etc to get meaningful information.
      */
-    void print(Serde<K> keySerde, Serde<V> valSerde);
+    void print(Serde<K> keySerde, Serde<V> valSerde, String streamName);
```
We should have 4 overloaded `print` methods, ie, this one should not be replaced but kept.

bbejeck replied:
Agreed.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -158,13 +158,19 @@ public KStreamImpl(KStreamBuilder topology, String name, Set<String> sourceNodes
     }
 
     @Override
-    public void print(Serde<K> keySerde, Serde<V> valSerde) {
```
keep this one and call `print(keySerde, valSerde, null)`

bbejeck replied:
Agreed

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -182,7 +188,7 @@ public void writeAsText(String filePath, Serde<K> keySerde, Serde<V> valSerde) {
 
             PrintStream printStream = new PrintStream(new FileOutputStream(filePath));
-            topology.addProcessor(name, new KeyValuePrinter<>(printStream, keySerde, valSerde), this.name);
+            topology.addProcessor(name, new KeyValuePrinter<>(printStream, keySerde, valSerde, null), this.name);
```
Should we extend (ie, new overloads) `writeAsText` with optional parameter `name` similar to `print()` ?

bbejeck replied:
> Should we extend (ie, new overloads) writeAsText with optional parameter name similar to print() ?

@mjsax 

I'm on the fence with this one.  I thought about adding the overloaded methods for `writeAsText()`, but decided not to.  The `print()` function writes to Stdout. With multiple streams the results are intermingled and steam/ktable names help with disambiguation.  The `writeAsText()` function writes the output of each stream/ktable to it's own file, so, in my opinion, having names for the stream/ktable doesn't add much.  

What do you think?

mjsax replied:
Makes sense. :)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java:
```diff
@@ -154,7 +154,7 @@ public void writeAsText(String filePath, Serde<K> keySerde, Serde<V> valSerde) {
         try {
             PrintStream printStream = new PrintStream(new FileOutputStream(filePath));
-            topology.addProcessor(name, new KeyValuePrinter<>(printStream, keySerde, valSerde), this.name);
+            topology.addProcessor(name, new KeyValuePrinter<>(printStream, keySerde, valSerde, null), this.name);
```
Same comment as above.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KeyValuePrinter.java:
```diff
@@ -69,11 +67,13 @@
             this.printStream = printStream;
             this.keySerde = keySerde;
             this.valueSerde = valueSerde;
+            this.streamName = (streamName == null) ? "" : "[" + streamName + "]: ";
```
Why replace `null` by empty string `""` ? If user specifies `null` we can just print `null`, IMHO.

bbejeck replied:
ak

mjsax commented:
Left some comments. Looks good in general.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -158,13 +158,19 @@ public KStreamImpl(KStreamBuilder topology, String name, Set<String> sourceNodes
+    public void print(Serde<K> keySerde, Serde<V> valSerde, String streamName) {
         String name = topology.newName(PRINTING_NAME);
-        topology.addProcessor(name, new KeyValuePrinter<>(keySerde, valSerde), this.name);
+        streamName = (streamName == null) ? this.name : streamName;
```
Hmm, I thought you planned to use the "parent processor node name" as the prefix if it is not specified, which seems more user friendly to me?

EDIT: actually my bad, nvm.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java:
```diff
@@ -137,7 +137,7 @@ public void print() {
     public void print(Serde<K> keySerde, Serde<V> valSerde) {
         String name = topology.newName(PRINTING_NAME);
-        topology.addProcessor(name, new KeyValuePrinter<>(keySerde, valSerde), this.name);
+        topology.addProcessor(name, new KeyValuePrinter<>(keySerde, valSerde, this.name), this.name);
```
Actually `this.name` is still the processor node name, not the KTable name as mentioned in the JIRA.

However, after thinking it a bit more, I feel there are a few corner cases when using the state store name may be cumbersome: even after KAFKA-3870 and KAFKA-3911 is merged, still not all KTables will be backed by a state store (for example, a KTable generated from another KTable.filter, which is just a view of the other table). And if we call `filter` on both of these KTables, they will actually share the same state store names, which are confusing. So just using the processor node name, admittedly are not very intuitive for users, may be the least bad solution here.

bbejeck replied:
With that in mind, I'll add the same overloaded methods for `print()` that were provided in `KStream`.

bbejeck committed:
KAFKA-3794: updates per review comments.

bbejeck commented:
@mjsax  @guozhangwang  changes per review comments:
- print null for stream name if null provided from user
- added additional overloaded print method with serdes but no stream name.
- added the same overloaded print methods to KTable that exist in KStream.

I found adding the stream name to `writeAsText` a little cumbersome.  How do you feel about using the parent node name for the stream name since there are not multiple streams to discern.

 Do you feel that this is a fair compromise?

mjsax commented:
With regard to you comment that the output of `writeAsText` goes to different files anyway, it might be better to have a "prefix-KeyValuePrinter" for `print()` only and use a non-prefix KeyValuePrinter for `writeAsText`. Just an idea.

bbejeck commented:
> With regard to you comment that the output of writeAsText goes to different files anyway..

@mjsax 

I prefer that approach, thanks for the input.

mjsax commented:
@guozhangwang Any input about `writeAsText` vs `print` ?

guozhangwang commented:
I think it does not hurt to still add the prefix as its parent stream name in `writeAsText`, although we let users to specify the file paths and in practice they may be different or maybe not. It's just like in our current log4j settings, where multiple classes could write to the same log file, but with different prefix still (as class names).

bbejeck commented:
Fair enough, adding streamName parameter to `writeAsText` as well.

bbejeck committed:
KAFKA-3794: updates per review comments, added streamName parameter to writeAsText methods

bbejeck commented:
@mjsax @guozhangwang 

Updates per latest reviews, added overloaded methods for `writeAsText` to specify the stream name for labeling key/value pairs.  At this point I believe all review comments have been addressed.

guozhangwang commented:
LGTM. Merged to trunk. Thanks @bbejeck !
